### ISSUE TITLE
Subcourse shows grades when they are hidden in the course

### ISSUE THREAD

**Issue by reactor0138** [mudrd8mz/moodle-mod_subcourse#28]
Looks like a wee bug. If we have the subcourse and its linked to a course where the grades are hidden for those students (not the entire final grade column) then the hidden results still appear in the subcourse. This is not good as it means the students have a back door to get to their final grades while we are still finalising them. I can't add pictures here so will add to moodle plug in

**Comment by mudrd8mz** [comment]
> not the entire final grade column

Do the students have access to the final grade in the referenced subcourse?

**Comment by reactor0138** [comment]
No if you you look at the referenced subcourse screen shot they have no access to the final grade but the subcourse activity ignores this and shows them the final grade

Many thanks

Scott Hallman

E-learning Co-ordinator

Extension 81254
03 3641254
Mobile 021 686 912

Monday 9-5
Tuesday 8:30 to 4:30
Wednesday 9-5
Thursday 8:30 to 4:30
Friday 8:30 to 4:30

**Comment by mudrd8mz** [comment]
Ok, I'll look at this. Meanwhile, if you are able to provide a patch, please feel encouraged to do so.

**Comment by reactor0138** [comment]
Thanks David

We’ll try to take a look at it too. Cheers

Many thanks

Scott Hallman

E-learning Co-ordinator

Extension 81254
03 3641254
Mobile 021 686 912

Monday 9-5
Tuesday 8:30 to 4:30
Wednesday 9-5
Thursday 8:30 to 4:30
Friday 8:30 to 4:30

**Comment by mudrd8mz** [comment]
BTW, how exactly do you have grades hidden in the referenced subcourse?

**Comment by mudrd8mz** [comment]
I am asking because the Subcourse module correctly checks the course setting "Show gradebook to students". So I assume you refer to the option when the course total grade item is hidden via the gradebook setup (and thence it does not appear at the student's Overview report), right?

**Comment by mudrd8mz** [comment]
I think this change will do it: https://github.com/mudrd8mz/moodle-mod_subcourse/compare/hidden-ref-grade 

It will help if you are able to test it and confirm it fixes your issue. Thanks in advance.

**Comment by reactor0138** [comment]
Hi David

Testing on 3.5 shows the patch did not work. That is if you amendment is in the zip download

The process I followed was.
Download the latest zip, and install via plugins zip upload
Add grades to feeder course
Add a subcourse linked to the feeder course in a base course
Check that the results appeared in the base course from the feeder course in the base course gradebook
Then hide the final grade in the feeder course via the hide icon in the grader report
Then go back to the base course and see if the grades still appear in the base course gradebook via the subcourse activity

Unfortunately the final grades still show via the subcourse activity despite it being hidden in the feeder course

Let me know if there is anything else I need to test or if you imagined that the grades would be hidden somewhere else

Cheers. Thanks for your time on this

Many thanks

Scott Hallman

E-learning Co-ordinator

Extension 81254
03 3641254
Mobile 021 686 912

Monday 9-5
Tuesday 8:30 to 4:30
Wednesday 9-5
Thursday 8:30 to 4:30
Friday 8:30 to 4:30

**Comment by reactor0138** [comment]
Hi David

Just for clarification there are 3 main ways to hide a final grade in a course.

Via grader report, turn on edting, then click the eye icon (this needs to be set up in graderreport setup to show) at the top of the column. This hides all final grades.
Via grader report, turn on edting, then click the eye icon, for a final grade of an individual student.

Via gradebook setup. Right click on the final grade item and chose hide

In essence all these options show then hide the final grade from coming through any subcourse activity, otherwise the students have an alternate way to get their grades even if we have hidden them. We hide their grades until the module or course has been totally finalised and then lock the grades

Many thanks

Scott Hallman

E-learning Co-ordinator

Extension 81254
03 3641254
Mobile 021 686 912

Monday 9-5
Tuesday 8:30 to 4:30
Wednesday 9-5
Thursday 8:30 to 4:30
Friday 8:30 to 4:30

**Comment by reactor0138** [comment]
Yes this is one way it can be hidden

Many thanks

Scott Hallman

E-learning Co-ordinator

Extension 81254
03 3641254
Mobile 021 686 912

Monday 9-5
Tuesday 8:30 to 4:30
Wednesday 9-5
Thursday 8:30 to 4:30
Friday 8:30 to 4:30

**Comment by mudrd8mz** [comment]
> That is if you amendment is in the zip download

No. This change is not yet present in the master branch. Either you can apply the change manually, or use download the ZIP generated off that feature branch: https://github.com/mudrd8mz/moodle-mod_subcourse/archive/hidden-ref-grade.zip

Please note that currently the patch deals with the case when the whole grade item (i.e. the whole column) is hidden in the referenced course. Hiding individual students' grades is not supported yet - and I actually haven't looked if that could ever be supported.

**Comment by mudrd8mz** [comment]
BTW can you please either comment via the Github interface, or make sure that your replies do not contain the footer. As you can see here at https://github.com/mudrd8mz/moodle-mod_subcourse/issues/28 the comments are flooded with your opening hours :-)

**Comment by mudrd8mz** [comment]
Have you had a chance to test the patch?

**Comment by reactor0138** [comment]
Hi David
sorry mate, I have been away for the last 10 days. I hope to get it tested this week. However it sounds like the patch will not fix the full issue, in that we would need it to work for all 3 hidden case scenarios. That is hidden in grades set up, hidden column in grader report and hidden for an individual student in the grader report. 
I hope to get back to you mid week
Cheers Scott

**Comment by mudrd8mz** [comment]
bq. and hidden for an individual student in the grader report.

As I suggested above, I don't currently think this is doable.

**Comment by reactor0138** [comment]
Hi David
we have tested your patch and the results are as follows. 
Hide via gradebook set up and hide final grade = fixed and working as expected
Hide via grader report and hide column (really same as above) = fixed and working as expected
Hide via grader report a single students final grade = no effect, final grade still comes though subcourse. final grades are leaked despite being hidden
Hide via grader report a group of students final grade = no effect, final grade still comes though subcourse. final grades are leaked despite being hidden
So it's working as you expected you patch would but does not work for sinlge and group of hidden final grades. We will have to hide the subcourse while grades are finalised until we/you can come up with a patch. Cheers

**Comment by mudrd8mz** [comment]
What seems to be doable is
```
diff --git a/locallib.php b/locallib.php
index 128c1451b4..b121a4a81d 100644
--- a/locallib.php
+++ b/locallib.php
@@ -127,6 +127,11 @@ function subcourse_fetch_refgrades($subcourseid, $refcourseid, $gradeitemonly =
                 continue;
             }
             $grade = new grade_grade(array('itemid' => $refgradeitem->id, 'userid' => $user->id));
+
+            if ($grade->hidden) {
+                continue;
+            }
+
             $grade->grade_item =& $refgradeitem;
             $return->grades[$user->id] = new stdClass();
             $return->grades[$user->id]->userid = $user->id;
```
so that if the grade is hidden in subcourse, it will not get copied over at all. Only after it is made available in the original subcourse, it would get copied. I quite like this behaviour and it would probably help in the described case, too.

**Comment by reactor0138** [comment]
Hi David
Thanks for looking into this as it is quite important to us and has been giving us lots of extra admin to keep grades hidden for a group. Do you have a working example that we can test? Or something we can download?

**Comment by mudrd8mz** [comment]
The patch above shows the three lines of code to be added to the locallib.php file to implement the behaviour.

**Comment by reactor0138** [comment]
Ok thanks We will try it out and let you know. Cheers

**Comment by mudrd8mz** [comment]
I finally found a solution for how to keep both hidden grade items (i.e. the whole column in the gradebook) and grades themselves (i.e. individual cells in the gradebook) synced as hidden. This supports the "hidden until" feature too. Just keep in mind that the subcourse activity module may be out of sync for a while so these updates may not be instant. The `grade_grade::set_hidden()` method does not trigger any observable event that we could react upon. So any change in the grade / grade item visibility must be either fetched explicitly (via the "Fetch now" link) or via the scheduled task (cron).

See 4b760c2 for details.

**Comment by donaldob** [comment]
**Related issue**
When final grades are hidden in a sub-course, the final grades in the underlying courses are not hidden.

I have modified view.php to enable the final grades in the underlying courses to be hidden, without having to visit each underlying course.

**Comment by clemoan** [comment]
I don't understand why this issue is closed. Perhaps I'm facing another issue or I don't understand subcource plugin behaviour. Subcourse displays total grade that takes in account hidden grades in the linked courses. So for students, the displayed grade is inconsistent or, worst, they can infer hidden grade.  
Could somebody explain what I miss, what is the expected behaviour ?
